### DISCUSSION TITLE
Proposal: Metadata DB

### DISCUSSION THREAD

**Discussion by thecodedrift** [avitorio/outstatic#53] [Ideas]
* [x] Proposal - This document, let's discuss
* [x] PoC - Proposal has linked proof of concept(s) showing the idea in action
* [x] Issue - Feature / Idea is ready, see linked issue and/or PR
* [x] Landed - Feature / Idea is landed and available

# Problem

_describe this in terms of something difficult to do in outstatic_

* Get all documents, sorted by publish date
* Get all documents in the X collection, written by Y
* Get all documents across all collections

The above expressions require a combination of `getDocuments`, `getDocumentPaths`, and `getCollections`; Implementing this inside of your own Next.js catch-all slug such as `/blog/[[...ost]].tsx` is an unwieldy requirement for even people familiar with Next.js. Additionally, to scan the entire document collection adds N fs reads and N yaml parses to `getStaticProps` which uses excessive compute time.

# Solution (Proposal)

**Metadata DB**. Taking cues from some of the original CMS systems such as [Movable Type](https://www.movabletype.org/), the cpu and file intensive operations should be done on write/commit instead of on read. In practice, this is because the number of times you write/edit a post is much smaller than the volume of reads coming into a CMS.

Whatever file solution we use, it would be a file in the root of outstatic (`<repo>/outstatic/content/` by default).

1. When a post is created or updated, the Metadata DB is loaded and the post's data by slug is changed/modified/added
2. The new Metadata DB file is added to the commit payload
3. Git/GitHub's conflict resolution ensures that simultaneous writes that clobber the DB file would fail since we [pass the `oid` through](https://github.com/avitorio/outstatic/blob/118c7ae6d8d8086183a7340d61f688079366d7f9/packages/outstatic/src/utils/createCommitInput.ts#L96)
4. A "Regnerate Metadata" button provides a UI based failsafe should something with metadata go horribly wrong. This button would regenerate the metadata file and make a single commit to update it with new data

## Solution Space

This is not a complete list of possible solutions, but instead designed to capture the range of ways we can approach a metadata file in-repo.

### [lowdb](https://github.com/typicode/lowdb) or just reading/writing JSON to disk

The strength of LowDB is its simplicity. At it's core, Low is a read/write abstraction that also is easy to extend for custom functionality. With a reasonable performance cap of 10MB worth of metadata, there's enough JSON to cover all the metadata needs in a very simple manner. The largest downside to Low is how little it gives you. Low offers no data access, requiring you to use lodash or a tool like [json-query](https://www.npmjs.com/package/json-query) or [filter-obj](https://github.com/sindresorhus/filter-obj) to perform any filtering. Even then, you would need to do your own ordering operation inside of `getStaticProps`.

Using something hyper-low level such as Low gives us the ability to craft our own query API if we so choose. For this example, I'm using json-query as the more powerful option, and then a usage of [thenBy](https://github.com/Teun/thenBy.js) which makes it possible to do a complex sort without losing your mind.

```ts
// interface for getting documents by a json-query chain
await getDocumentsBy(posts => posts
  .filter("metadata[category=birds]")
  .filter("metadata[collection=posts]")
  .sort(firstBy("metadata.publishedAt))
  .limit(10)
)
```

Alternative query systems: [jsonpath](https://github.com/dchester/jsonpath), [node-jq](https://github.com/sanack/node-jq)

### [LokiJS](https://github.com/techfort/LokiJS) or other JavaScript DB solutions

Take something like Low, add a query and chaining API, add binary indexes, and you get Loki. It's performance is similar to Low, as both write/read JSON under the hood. Loki's binary indexes are included in its export via `serialize()`. The additional JSON structure increases the chances of a merge conflict if two people are updating metadata at the same time, though this should be caught by our `oid` checking code as part of `createCommitInput`.

Loki has two major downsides: The project was last updated in March, though we don't need anything [on the roadmap](https://github.com/techfort/LokiJS#roadmap), and second there is an [absurd number of forks](https://github.com/techfort/LokiJS/network).

Code using loki would look something like (using [query examples](https://rawgit.com/techfort/LokiJS/master/docs/tutorial-Query%20Examples.html))

```ts
// get documents by a loki metadata query
await getDocumentsBy(posts => posts
  .chain()
  .find({
    "metadata.category": "birds",
    "metadata.collection": "posts"
  })
  .simplesort("metadata.publishedAt")
  .limit(10)
);
```

Other js + db solutions: [PouchDB](https://pouchdb.com/), [teDB](https://github.com/tedb-org/teDB), [Watermelon](https://nozbe.github.io/WatermelonDB/), [warehouse](https://github.com/hexojs/warehouse)

### [sql.js](https://github.com/sql-js/sql.js) or another SQL on disk system

At the complex end of the spectrum is something like SQLite, of which sql.js is a wasm version of the C code and can run completely in JavaScript. It's a heavyweight answer to what we're trying to solve with metadata, but does give us access to well known tools when it comes to things like data migration. The biggest downside is that we're asking every outstatic user to learn SQL for querying metadata.

```ts
await getDocuments(`
  SELECT slug from metadata
  WHERE category = "birds" AND collection = "posts"
  ORDER BY publishedAt DESC
  LIMIT 10
`);
```

# Recommendations

1. `getDocumentBySlug` will remain as the easiest way to do your outstatic data. For more powerful implementations, a lower level function will be available, tentatively named `getDocumentBy`
2. As a file system (and git) based CMS, the saved file(s) should be readable and editable on GitHub. Otherwise, we're undermining one of outstatic's strengths
3. Metadata is extremely author-dependent. That means any DB with typescript support will probably end up typing metadata as `Record<string, unknown>`. In _theory_ a types file could be generated from your custom metadata `schema.json` but that's beyond the scope of what we're doing here.

# References

* https://github.com/avitorio/outstatic/discussions/28#discussioncomment-4072271

**Comment by avitorio** [comment]
Excellent write up and summary of our current status @jakobo. Thank you!

These suggestions are great. I will take some time this week to read about, and play with, the proposed solutions.

**Reply by thecodedrift** [reply]
Sounds good. I did some playing around as well and updated my findings and recommendations. If you can think of anyone we should /cc on this, do tag them.

**Reply by thecodedrift** [reply]
Further updates added RE experimenting with Pouch (which is back to active status after a hiatus).

**Comment by thecodedrift** [comment]
I think full-on SQL is overkill at this point. I also think asking users to bring their own JSON-filtering-tool goes against out desire to make outstatic accessible to as many people as possible.

A JS document w/ query support is probably the most straightforward. Powerful features like query/sort/limit being built-in will reduce barrier to entry for moderate-to-complex cases such as tagged posts or all posts by an author, sorted by time, and paginated.

In order, I'd suggest LowDB +  [mingo](https://github.com/kofrasa/mingo), and if the size of the JSON document begins to impact performance, we explore either Loki or PouchDB. The data can be stored in `/outstatic/content/metadata.json`, and contains both hash information about the file's computation as well as our metadata entries.

```json5
{
  "hash": "abcdef08" // represents the hash from which this metadata db is built and should match HEAD~1 on checks
  "data": [
    {  // sample post object
      "collection": "posts",
      "slug": "a-sample-slug",
      "hash": "ffffff08", // sha-1 of when the file entry
      "metadata": {
        // translate YAML into key-value pairs direct from metadata
      }
    }
  ]
}
```

The `hash` keyword allows us to determine during `getStaticProps` / `getServerSideProps` what part(s) of the Metadata DB need to be rebuilt. The top level hash represents `HEAD~1` of the `/outstatic/content` directory. On mismatch, the contents of the outstatic content directory were changed, and the DB needs to be reconciled. During reconciliation, we return dirty data.

On reconciliation, each individual file can be checked against its entry in the Metadata DB. If we get a match (slug=x hash=y), then the metadata is already up to date and that file can be skipped. Otherwise we load the file and its git rev, parse the YAML, and upsert the metadata record. When the files are reconciled, the top level hash can be set and committed, triggering a next.js rebuild.

Ideally, rebuilds should be rare. On editing and deleting a file, we should include the metadata DB changes as part of the commit changeset. Our oid checks already prevent attempting a save on a stale HEAD.

# Build Plan

* Phase 1: a metadata.json in `/outstatic/content` that can be built by reading the outstatic directory off of github and doing a full reconciliation, button in UI to perform this operation
* Phase 2: a getDocumentsBy that exposes a metadata query system for users/developers
* Phase 3: hook into save/delete steps to make changes to metadata.json
* Phase 4: SWR integration with getDocumentsBy, doing a background reconcilliation

**Comment by revskill10** [comment]


**Reply by avitorio** [reply]
Do you mean on the Dashboar/Editor? We are working on it.

**Reply by revskill10** [reply]
Hi @avitorio , how about adding collaborators through github ? I tried add a collaborator as admin, but after signing in via github, he still can't adding content.

**Reply by avitorio** [reply]
Hey @revskill10, I invite you to join our [discord channel](https://discord.gg/qEjtpn7E6F) to continue this discussion. As this thread should be about the Metadata DB work.

**Comment by thecodedrift** [comment]
## Exposing the Query Language (Query API)

This is just some spitballing about how the "query" function might work inside of `getStaticProps` (gsp). Currently, the example gsp looks like:

```ts
type Params = {
  params: {
    slug: string
  }
}

export async function getStaticProps({ params }: Params) {
  const post = getDocumentBySlug('posts', params.slug, [
    'title',
    'publishedAt',
    'slug',
    'author',
    'content',
    'coverImage'
  ])
  const content = await markdownToHtml(post.content || '')

  return {
    props: {
      post: {
        ...post,
        content
      }
    }
  }
}
```

There's a couple things outstatic does really well here. First, converting the content (markdown) is left up to the end user, letting them select MDX, markdoc, or plain markdown via remark. Second, there's a single API that uses the file system to fetch the content. With a metadata DB, we need to expose a query-like API against the db which can return our requested fields.

This would be the proposed query API, exposed as `{ query }` in `outstatic/server`

```ts
import { query } from 'outstatic/server'

export async function getStaticProps({ params }: Params) {
  const post = query({
    category: "posts",
    slug: params.slug
  }, {
    title: 1,
    publishedAt: 1,
    slug: 1,
    author: 1,
    content: 1, // under the hood, this is added in via outstatic, even though it's from the fs and not metadata
    coverImage: 1
  }).next() // takes first item off collection

  const content = await markdownToHtml(post?.content ?? "");

  return {
    props: {
      post: {
        ...post,
        content,
      },
    },
  };
}
```

This looks a lot like mongo query/projection because mingo, sift, couch, and others all opted for mongo-like syntax. I agree with the general consensus; the JSON model is easy to reason about.

**Reply by avitorio** [reply]
Excellent. Having a query function will definitely make the lib a lot more powerful. Thank you!

**Comment by avitorio** [comment]
The Metadata DB has been added to the project! Thanks @jakobo
